rfd/0058-desktop-file-transfer.md

```diff
@@ -0,0 +1,143 @@
+
+This clever system makes client implementations easier, because the client isn't responsible for monitoring and alerting the server of changes to the shared files. For example, if a user edits and writes to a shared file on the client side in another program that the RDP client doesn't control (like a text editor), the client doesn't need to recognize the change and communicate it back to the server. Instead, it just sends whatever file data is on disk if/when the server asks for it, and so those sorts of situations are handled without such low level control. Additionally, this system minimizes bandwidth by only sending whatever data is actually needed, when its needed (from client to server). For example, when a directory is first shared, the server only asks for metadata like the file names and sizes in the directory, which it can then display to the user as if those files are living on the remote Windows system. Only when a user actually opens one of those files does it request the file data itself. Also, if the file is a gigantic text file that the user opens in a well optimized text editor, it needn't request the entire file at once. Instead, it can just request the section of the file needed to be displayed at that time. There is no file size limit with this option.
+
+For our purposes, the file system virtual channel extension option (RDP Option 2) can be broken down into two sub options, RDP Option 2a and RDP Option 2b. RDP Option 2a is that we share a directory on the machine running the RDP client (`windows_desktop_service`), and RDP Option 2b is sharing a directory that lives on the Teleport user's client itself (via the File System Access API, discussed in more detail below).
+
+## Client Side Possibilities
+
+Irrespective of which option we choose, we will need some means of initiating and completing file transfers on the client side (from the browser).
+
+### Clipboard API
+
+At first glance, it seemed like using the `Clipboard.read` and `Clipboard.write` API's might be a possibile mechanism for us to make use of. However while some files can be transferred to the client's clipboard with this API, browsers are [constrained](https://w3c.github.io/clipboard-apis/#reading-from-clipboard) to the `text/plain` `text/html` and `image/png` mime types. Ergo unless we want to limit our users to a very narrow set of file transfers, this option is not available to us.
+
+### Browser Upload and Download (Client Option 1)
```

Comment:
I'm having a hard time keeping track of which option is being discussed due to the deep level of nesting here.

Is "Client Option 1" a sub-option of RDP Option 2a?

Reply:
Yeah, my naming system is questionable, in hindsight the numbering might be more confusing than helpful. I tried to clarify by the structure of the headings, but its difficult to follow, particularly reading the plain Markdown. 

I've done some re-naming and here is a summary of the structure based on the headings:
- **RDP Possibilities and Data Flow (RDP Options 1, 2a, and 2b)**
  - Clipboard Virtual Channel Extension (RDP Option 1)
  - File System Virtual Channel Extension (RDP Option 2a and 2b)
- **Client Side Possibilities (Client Options 1 and 2)**
  - Browser Upload and Download (Client Option 1)
  - TDP Upload and Download via the File System Access API (Client Option 2)
- **Implementation Options (Full Implementation Options 1, 2, 3, and 4)**
  - RDP Option 1 + Client Option 1 (Full Implementation Option 1)
  - RDP Option 1 + Client Option 2 (Full Implementation Option 2)
  - RDP Option 2a + Client Option 1 (Full Implementation Option 3)
  - RDP Option 2b + Client Option 2 (Full Implementation Option 4)

The first two **bolded bullet points** contain our options on the backend and frontend respectively, the last one lays out the plausible combinations of first two that are worth considering.